torch/distributed/algorithms/quantization.py

```diff
@@ -0,0 +1,130 @@
+
+
+from enum import Enum
+
+
+TORCH_HALF_MIN = torch.finfo(torch.float16).min
+TORCH_HALF_MAX = torch.finfo(torch.float16).max
+
+class QuantType(Enum):
+    FP16 = "fp16"
+
+    def __str__(self) -> str:
+        return self.value
+
+
+def _fp32_to_fp16_with_clamp(tensor: torch.Tensor) -> torch.Tensor:
+    return torch.clamp(tensor, TORCH_HALF_MIN, TORCH_HALF_MAX).half()
+
+def _quantize_tensor(tensor, qtype):
+    if not isinstance(tensor, torch.Tensor):
+        raise RuntimeError(
+            "Invalid function argument. Expected tensor "
+            "to be a single tensor list  of type torch.Tensor."
+        )
+    if (qtype == QuantType.FP16):
+        return _fp32_to_fp16_with_clamp(tensor)
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+def _quantize_tensor_list(tensor_list, qtype):
+    if not isinstance(tensor_list, list) or not all(
+        isinstance(p, torch.Tensor) for p in tensor_list
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be of type List[torch.Tensor]."
+        )
+    if (qtype == QuantType.FP16):
+        input_tensor = [_quantize_tensor(t, qtype) for t in tensor_list]
+        return input_tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+def _dequantize_tensor(tensor, qtype, quant_loss=None):
+    if not isinstance(tensor, torch.Tensor):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be a single tensor list  of type torch.Tensor."
+        )
+    if (qtype == QuantType.FP16):
+        if tensor.dtype != torch.float16:
+            raise RuntimeError(
+                "tensor dtype is {} while expected to be FP16.".format(
+                    tensor.dtype
+                )
+            )
+        elif tensor.dtype == torch.float16 and quant_loss is None:
+            return tensor.float()
+        else:
+            return tensor.float() / quant_loss
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+
+def _dequantize_tensor_list(tensors, qtype, quant_loss=None):
+    if not isinstance(tensors, list) or not all(
+        isinstance(p, torch.Tensor) for p in tensors
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be of type List[torch.Tensor]."
+        )
+    elif (qtype == QuantType.FP16):
+        tensor = [_dequantize_tensor(t, qtype) for t in tensors]
+        return tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+
+def auto_quantize(func, qtype, quant_loss=None):
+    """
+    This is a prototype API that automahically quant  the input tensors, choose the precision types, and
+    pass other necessary arguments and then dequantizes the output.
```

Comment:
It's beneficial to document the argument and return types so that user of this API can understand how to use it, can you follow documentation principles in `distributed_c10d.py` to document argument and return types?